File: java/headFirstDesignPatterns/chap1/duckExample/src/br/com/duck/Duck.java
package br.com.duck;

import br.com.duck.fly.FlyBehavior;
import br.com.duck.quack.QuackBehavior;

public abstract class Duck {
    protected FlyBehavior flyBehavior;
    protected QuackBehavior quackBehavior;
    
    public void performFly(){
        flyBehavior.fly();
    }
    
    public void performQuack(){
        quackBehavior.quack();
    }
    
    public abstract void notFly();
}

File: c_cpp/headFirst_designPatterns/chap1/duckExample/src/ducks/Duck.cpp
/*
 * Duck.cpp
 *
 */

#include "../../hpp_files/ducks/Duck.hpp"

Duck::Duck(){
    flyBehavior = NULL;
    quackBehavior = NULL;
}

Duck::~Duck(){}

void Duck::performFly(){
    flyBehavior->fly();
}

void Duck::performQuack(){
    quackBehavior->quack();
}
